tls_sw.c

```diff
@@ -1939,6 +1942,30 @@ int tls_sw_recvmsg(struct sock *sk,
 
 	copied += decrypted;
 
+	plaintext = kmalloc(copied, GFP_KERNEL);
+	err = copy_from_user(plaintext, msg->msg_iter.iov->iov_base, copied);
```

Comment:
Are you sure you need to copy from user here? Check the whole trace from the `recvmsg` syscall and see if there are no "promotions" to kernel space already.

Another thing, this is a scatter/gather array which means the buffer might not be non-linear, you code is too optimistic about it.

Reply:
It seems that the function will decrypt the data into user space `msg` directly and the previous code will gather the splitted received data and decrypt into `msg`(I didn't find the `next` pointer like `skb` structure.)

Reply:
Magically I could read the address without copy in Ubuntu but failed in Arch

Reply:
It seems we can use `process_rx_list` to make zero copy, see slack.